### DISCUSSION TITLE
how to get  transportOrder state in the vehicle driver

### DISCUSSION THREAD

**Discussion by scsq122** [openTCS/opentcs#11] [Q&A] [answered]
hi, I am trying to get transportOrder state(BEING_PROCESSED、WITHDRAWN、FINISHED...) in the vehicle driver, what should I
 do? In Vehicle.java, I find this code:
  /**
   * A reference to the transport order this vehicle is currently processing.
   */
  private final TCSObjectReference<TransportOrder> transportOrder;
but I do not know how to get the transportOrder state in the vehicle dirver. thanks.

**Comment by swltr** [comment, marked as answer]
Hi,

in your `VehicleCommAdapter` implementation, use an injected instance of `TransportOrderService` to fetch the `TransportOrder` instance. (Also see [the Developer's Guide](https://www.opentcs.org/docs/5.3/developer/developers-guide/opentcs-developers-guide.html#section-kernel-api) on this.) Actually, for fetching a `TCSObject` like `TransportOrder`, a `TCSObjectService` is also sufficient.

What do you need the transport order's state for, though? If it's the transport order the vehicle is currently processing, the state will always be `BEING_PROCESSED`...

Best regards,
Stefan

**Reply by scsq122** [reply]
thanks for your reply, i want to do some options in vehicle driver when the transportOrder withdrawed.  this is
why i want to get the transportOrder state in the vehicle driver. when the transportOrder withdrawed, can I get
this "withdrawed event" in other ways?

**Reply by swltr** [reply]
OK. The vehicle driver does not receive an explicit notification when a transport order has been withdrawn from the vehicle. (It simply does not receive any further movement commands for the withdrawn transport order and will eventually receive movement commands belonging to another transport order.)

Using a `TransportOrderService` or `TCSObjectService` as described will work, with the implication that you'll have to poll and check for changes. An alternative would be to subscribe with the kernel's event bus (see [the Developer's Guide](https://www.opentcs.org/docs/5.3/developer/developers-guide/opentcs-developers-guide.html#_using_the_event_bus)) and filter for instances of `TCSObjectEvent` related to the transport order. That eliminates the polling, but also means you'd get a lot of events you don't care about.

**Reply by scsq122** [reply]
thanks，I see， I will try your method。thank you very much。
